plugins/visual-qontract/src/components/CodeComponentsComponent/CodeComponentsComponent.tsx

```diff
@@ -18,14 +17,61 @@ import {
     InfoCard,
 } from '@backstage/core-components';
 import { CodeComponentsQuery } from './query';
+import { RELATION_HAS_PART } from '@backstage/catalog-model';
+import { useEntity, useRelatedEntities } from '@backstage/plugin-catalog-react';
 import QueryQontract from '../../common/QueryAppInterface';
 
 
 export const CodeComponentsComponent = () => {
+    const title = "Code Repositories"
 
-    const { result, loaded, error } = QueryQontract(CodeComponentsQuery)
+    const { entity } = useEntity();
 
-    const title = "Code Repositories"
+    const { entities: relatedEntities, loading: relatedEntitiesLoading, error: relatedEntitiesError } = useRelatedEntities(entity, {
+        type: RELATION_HAS_PART,
+        kind: "Component"
+    });
+
+    const { result: qontractResult, loaded: qontractLoaded, error: qontractError } = QueryQontract(CodeComponentsQuery);
+
+    if (relatedEntitiesError || qontractError) {
+        return <InfoCard title={title}>
+            <Typography align="center" variant="body1">
+                Error loading the {title} information.
+            </Typography>
+        </InfoCard>
+    }
+
+    if (relatedEntitiesLoading || !qontractLoaded) {
+        return <InfoCard title={title}>
+            <Typography align="center" variant="body1">
+                Loading...
+            </Typography>
+        </InfoCard>
+    }
+
+    if (qontractResult.apps_v1.length === 0 || qontractResult.apps_v1[0].codeComponents.length === 0) {
+        return (
+          <InfoCard title={title}>
+            <Typography align="center" variant="body1">No {title} found.</Typography>
+          </InfoCard>
+        );
+    }
+
+    // For each code component, try to locate a child entity which has a matching source code URL.
+    // If a match is found, use that entity's 'image-build-url' annotation value if defined
+    for (var component of qontractResult.apps_v1[0].codeComponents) {
```

Comment:
You'd want to put this code into a side effect rather than have it sitting here in the body of component. The reason is that any code in the body of the component will re-run every time React triggers a render. This can lead to code being run at time you don't expect, or even worse it can result in render bombs where code is run on render triggers a render and so on. In this case your code is pretty safe. You aren't producing a side effect from this code so it wont trigger a render, and the code is pretty light so even if it re-runs a bunch of times it wont really hurt anything - and the component itself is simple enough that renders likely wont occur except on mount. However, I want to point this out and change it because this kind of thing can lead to really hard to understand issues in more complex cases. Since we're all getting up to speed on this stuff together I think it is best to try and learn the right patterns now so we don't have to unlearn them later.

I suggest you refactor this to put the for loop in a side effect that triggers on changes to `qontractResult` and `relatedEntities`:

```
    useEffect(() => {
        // For each code component, try to locate a child entity which has a matching source code URL.
        // If a match is found, use that entity's 'image-build-url' annotation value if defined
        for (var component of qontractResult.apps_v1[0].codeComponents) {
            if (relatedEntities) {
                for (var relatedEntity of relatedEntities) {
                    if (relatedEntity.metadata.annotations?.["backstage.io/source-location"] == `url:${component.url}`) {
                        var annotationBuildUrl = relatedEntity.metadata.annotations?.["visual-qontract/image-build-url"]
                        if (annotationBuildUrl) {
                            component.imageBuildUrl = annotationBuildUrl
                        }
                    }
                }
            }
        }
    }, [relatedEntities, qontractResult])
```

This will ensure that your code only runs when either of those state variables change and not on render.